src/Razor/src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs

```diff
@@ -208,48 +212,129 @@ internal static bool TryComputeNamespaceAndClass(this RazorCodeDocument document
 
             var filePath = document.Source.FilePath;
             var relativePath = document.Source.RelativePath;
-            if (filePath == null || relativePath == null || filePath.Length <= relativePath.Length)
+            if (filePath == null || relativePath == null || filePath.Length < relativePath.Length)
             {
                 @namespace = null;
-                @class = null;
                 return false;
             }
 
-            filePath = NormalizePath(filePath);
             relativePath = NormalizePath(relativePath);
-            var options = document.GetCodeGenerationOptions() ?? document.GetDocumentIntermediateNode()?.Options;
-            var rootNamespace = options?.RootNamespace;
-            if (string.IsNullOrEmpty(rootNamespace))
+            // If the document or it's imports contains a @namespace directive, we want to use that over the root namespace.
+            var baseNamespace = string.Empty;
+            var appendSuffix = true;
+            var lastNamespaceContent = string.Empty;
+            var lastNamespaceLocation = SourceSpan.Undefined;
+            var importSyntaxTrees = document.GetImportSyntaxTrees();
```

Comment:
Should this throw if parsing isn't done.

Reply:
That's debatable. We could still try to compute it using the RootNamespace and the file path. But the counter argument could be that if there is a `@namespace` that we haven't parsed yet then the namespace we compute could potentially be wrong.

Reply:
I don't think it's very reasonable to try and behave half correct. I think throwing is best.